docs/interpreter_design_doc.md

```diff
@@ -0,0 +1,100 @@
+# Interpreter Design
+
+
+## Class Hierarchy
+
+Stablehlo programs are computations over Tensors (n-dimensional arrays), hence
+most of the input/output data values for an op are Tensors with the exception of
+constant
+[Attribute](https://docs.google.com/document/d/1GAAK_yHk7q2cZdJ3zBhsiNcrqzfpyjwQ-oWLD7q__Ss/edit#heading=h.5nem09i7p9va)
+values of primitive types (e.g., Int, float etc.).
+
+In the current design, the class `Tensor` implements Tensor data. The underlying
+storage class for `Tensor` objects, `detail:Buffer`, stores the type of the
+Tensor along with a contiguous byte array representing its data.
+
+Individual elements of a Tensor are represented using `Element` class which uses
+`mlir::Attribute` for storage.
```

Comment:
"Using `mlir::Attribute` simplifies things because this means that we don't have to implement our own machinery for storing values of different types inside `Element`".